python/util.py

```diff
@@ -0,0 +1,257 @@
+from functools import wraps
+from collections import namedtuple
+from itertools import starmap
+from operator import add, sub, eq
+
+
+def cache(method):
+    attr_name = '_' + method.__name__
+    @wraps(method)
+    def new_method(self):
+        if not hasattr(self, attr_name):
+            setattr(self, attr_name, method(self))
+        return getattr(self, attr_name)
+    return new_method
+
+
+def overlay_lists(base, new, offset):
+    '''Mutates the given base list by overlaying the new list on top at a given
+    offset.
+    '''
+    if offset + len(new) >= 0:
+        for i in range(len(new)):
+            base_idx = i + offset
+            if 0 <= base_idx < len(base):
+                base[base_idx] = new[i]
+            elif base_idx >= len(base):
+                break
+
+
+class Clamped:
+    '''Descriptor that limits the range of an attribute, given two callbacks
+    values or callbacks for minimum values.
+    '''
+    def __init__(self, min_, max_):
+        self.min = min_ if callable(min_) else lambda _: min_
+        self.max = max_ if callable(max_) else lambda _: max_
+
+    @property
+    def identifier(self):
+        return '_{:x}'.format(id(self))
+
+    def __get__(self, obj, type_):
+        if obj is None:
+            return self
+        else:
+            return getattr(obj, self.identifier)
+
+    @staticmethod
+    def _limit(value, min_, max_):
+        if min_ is None and max_ is None:
+            return value
+        if min_ is None:
+            min_ = min(value, max_)
+        if max_ is None:
+            max_ = max(value, min_)
+        return sorted((value, min_, max_))[1]
+
+    def __set__(self, obj, value):
+        min_ = self.min(obj)
+        max_ = self.max(obj)
+        setattr(obj, self.identifier, self._limit(value, min_, max_))
+
+UNASSIGNED = object()
+
+
+class AB:
+    '''Vector like class to help process multiple related values together.
+    Provides useful functionality like mapped attribute lookups and
+    distributive calling.
+    '''
+
+    __slots__ = '_data',
+
+    def __init__(self, a, b=UNASSIGNED):
+        'Takes either values for a and b or a value to repeat for both'
```

Comment:
Yes please, I know they're not required but it just doesn't look like a docstring without them.